third_party/pcsc-lite/naclport/server_clients_management/src/client_request_processor.cc

```diff
@@ -410,33 +437,118 @@ GenericRequestResult PcscLiteClientRequestProcessor::SCardConnect(
+  return return_code;
+}
+
+LONG PcscLiteClientRequestProcessor::SCardConnectHandleProtocolMismatchError(
+    SCARDCONTEXT s_card_context,
+    const std::string& reader_name,
+    DWORD share_mode,
+    DWORD preferred_protocols,
+    LPSCARDHANDLE s_card_handle,
+    LPDWORD active_protocol) {
+  LONG return_code = SCARD_E_PROTO_MISMATCH;
+  // If SCardConnect fails with SCARD_E_PROTO_MISMATCH it could be because a
+  // client application did not correctly reset the card and is now trying to
+  // reuse the previous connection although it requires a different protocol. If
+  // allowed by admin policy, disconnect and reset the card, and retry
+  // connecting to it.
+
+  optional<AdminPolicy> admin_policy = admin_policy_getter_->WaitAndGet();
+  std::vector<std::string> client_ids;
+  if (admin_policy &&
+      admin_policy.value().scard_disconnect_fallback_client_app_ids) {
+    client_ids =
+        admin_policy.value().scard_disconnect_fallback_client_app_ids.value();
+  }
+
+  if (std::find(client_ids.begin(), client_ids.end(),
+                this->client_name_for_log_) != client_ids.end()) {
+    GOOGLE_SMART_CARD_LOG_INFO
+        << logging_prefix_ << "SCardConnect failed with a protocol mismatch "
+        << "error. Attempting SCardDisconnect fallback: disconnecting and "
+        << "resetting any previous connections for context "
+        << DebugDumpSCardContext(s_card_context);
+
+    bool reset_successful =
+        SCardConnectAttemptReset(s_card_context, reader_name, share_mode);
+    if (reset_successful) {
+      GOOGLE_SMART_CARD_LOG_DEBUG
+          << logging_prefix_ << "Card reset finished successfully. Retrying "
+          << "to establish the requested SCardConnect connection ...";
+
+      return_code =
+          ::SCardConnect(s_card_context, reader_name.c_str(), share_mode,
+                         preferred_protocols, s_card_handle, active_protocol);
+    }
+
+    if (return_code != SCARD_S_SUCCESS) {
+      GOOGLE_SMART_CARD_LOG_INFO
+          << logging_prefix_ << "SCardDisconnect fallback was unsuccessful.";
+    }
+  }
+
+  return return_code;
+}
+
+bool PcscLiteClientRequestProcessor::SCardConnectAttemptReset(
+    SCARDCONTEXT s_card_context,
+    const std::string& reader_name,
+    DWORD share_mode) {
+  // Try to get an s_card_handle by performing SCardConnect using any protocol
+  // and without checking or updating the card handle registry. Do not handle
+  // potential protocol mismatch errors since at this point the reset attempt is
+  // already doing that.
+  DWORD preferred_protocols =
+      SCARD_PROTOCOL_ANY | SCARD_PROTOCOL_RAW | SCARD_PROTOCOL_T15;
+  SCARDHANDLE s_card_handle;
+  DWORD active_protocol;
+  LONG return_code_connect =
+      SCardConnect(s_card_context, reader_name, share_mode, preferred_protocols,
+                   &s_card_handle, &active_protocol,
+                   /*update_card_handles_registry=*/false,
+                   /*handle_protocol_mismatch_error=*/false);
```

Comment:
This recursive call makes the code a bit hard to follow. IIUC, the `handle_protocol_mismatch_error` parameter is also there to prevent infinite recursion here. I also don't like that we lost some checks/calls (like `OnSCardContextRevoked()`) in the new workaround's code path. What if we restructure the code a bit and streamline it?

```
LONG SCardConnect(...) {
  FunctionCallTracer ...
  LONG return_code = ObtainCardHandleWithFallback(...);
  tracer.LogExit();
  return return_code;
}

LONG ObtainCardHandleWithFallback(...) {
  LONG return_code = ObtainCardHandle(...);
  if (return_code == SCARD_E_PROTO_MISMATCH && IsDisconnectFallbackPolicyEnabled()) {
    if (ResetCard(...) && ObtainCardHandle(...)) {
      return SCARD_S_SUCCESS;
    }
  }
  return return_code;
}

LONG ObtainCardHandle(...) {
  LONG return_code = ::SCardConnect(...);
  if (return_code == SCARD_S_SUCCESS)
    s_card_handles_registry_.AddHandle(...);

  // Catch when PC/SC-Lite core, unlike us, thinks the context doesn't exist.
  if (return_code == SCARD_E_INVALID_HANDLE)
    OnSCardContextRevoked(s_card_context);

  return return_code;
}

bool ResetCard(...) {
  if (ObtainCardHandle(...) != SCARD_S_SUCCESS) {
    return false;
  }
  return SCardDisconnect(...) == SCARD_S_SUCCESS;
}
```

Would something like this work?